# Ð¡ommand-line parameters

	
**POST-INSTALL EXAMPLE DATA**

- `--sampledata`  
Creates the 'svist4get' folder in the current working directory.
The folder will contain sample data sets and adjustable configuration file templates.


**MANDATORY ARGUMENTS**

- `-t` <id>  
Visualization of the genomic window of a particular
transcript (using GTF transcript_id).

	OR

- `-g` <id>  
Visualization of the genomic window of a particular
gene (using GTF gene_id).

	OR

- `-t` <id> `-w` tss|tis <upstream> <downstream>  
Visualization of the genomic window centered on a transcription start
site|translation initiation site of a particular transcript (using GFF transcript_id) with a fixed offsets upstream and downstream.

	OR 

- `-w` <contig> <start> <end>  
 Visualization an arbitrary genomic window using genomic coordinates.


- `-fa` <file.fa>  
Path to the genome assembly multifasta for the sequence track.  
NOTE: the multifasta file must contain all contigs in a single file.  
NOTE: fai index file will be generated automatically.


- `-gtf` <file.gtf>  
Path to the gtf file with the transcript annotation.


**OPTIONAL ARGUMENTS**

- `-bg` <file1.bedGraph> [<file2.bedGraph>...]  
Paths to bedGraph files for genomic signal tracks.

- `-pbg` <file1A.bedGraph file1B.bedGraph> [`-pbg` <file2A.bedGraph file2B.bedGraph>...]  
Paths to 2 bedGraph files to be used for the paired genomic signal track (e.g. to plot signals for + and - DNA strands together). Positive and negative range of the Y-axis will be used for the first and second file, respectively (or vice versa if the reverse complementary transformation using -rc parameter is requested). The track will use absolute values from each bedGraph. Multiple paired tracks can be generated by using multiple '-pbg' parameters.  
For this track, the signs of values in each bedGraph will be ignored.


- `-c` <path> |A4_p1|A4_p2|A4_l  
Path to a configuration file or name of the premade config file (can be A4_p1, A4_p2, A4_l).


- `-o` <name>  
Output file name. Generates a pdf (vector) or png (raster) file based on the provided extension. Both png and pdf will be generated if the extension is not specified.


- `-hi`  
Hide intronic segments (default: false).


- `-rc`  
 Reverse-complement transformation of the genomic window (default: off).


- `-bgb` mean|median|max|min|none  
By default, the bedGraph tracks are plotted with 'bars' spanning small genomic windows of automatically optimized length. This parameter sets the function to aggregate bedGraph signal values for a single bar of the bedGraph tracks.  Default: 'mean'. Alternatives: 'max', 'min', 'median' and 'none'. 'none' forces plotting the raw signal at 1nt resolution.


- `-bg_log`  
Logarithmic (log2) scale for the each bedGraph (Y-axis)



- `-nts`  
Show reference genomic sequence with single-nucleotide resolution (default: hide).


- `-aas` auto|tics|codons  
Style of the aminoacid sequence track: tics (suitable for long regions e.g. > 160 nts, shows only start and stop codons with colored tics) or codons (show each aminoacid as labeled block) (default: auto).


- `-it` <text>  
Image title.


- `-xs` <N>  
 X axis tics step (nt) (default: auto).


- `-lb` <N1> [<N2>..]  
Y axis tics levels for a bedGraph track, a space-separated list of values, a separate value for each bedGraph track (default: auto).


- `-ys` <N1> [<N2>...]  
Step between axis tics a the bedgraph track Y-axis, a space-separated list of values, a separate value for each bedGraph track (default: auto).


- `-bul` <N1> <N2>... | max  
Upper limit for a bedGraph track Y-axis, can be  
 1. a space-separated list of values: a separate axis limit for each bedGraph track.  
 2. max: a maximum reachable value across all visible tracks.  
Default: reachable maximum for each track separately.


- `-bll` <N1> <N2>... | min  
Lower limit for a bedGraph track Y-axis, can be  
 1. a space-separated list of values, a separate value required for each bedGraph track.  
 2. min: a minimum reachable value across all visible tracks.  
Paired bedGraph uses absolute values for both signal tracks, since the negative Y-axis range is used to display the 2nd track.


- `-bl` <tex1t> <text2>  
Text labels for the bedGraph tracks, space-separated list of values, a separate value required for each bedGraph track (default: names of the bedGraph files).

- `-blp` left|center|right  
Text label position for the bedGraph tracks (default: center)


- `-bgc` <color1> [<color2>]  
Space-separated list of colors for bedGraph tracks. The colors should be specified by names listed in the palette file (default: internal palette or palettes/palette.txt). The list will be cycled through if shorter than the number of bedGraph tracks. Paired and basic bedGraph tracks use joint color list.


- `-gi` <s-e> [<s1-e1>...]  
Additional track showing selected genomic intervals as segments, space-separated list of start-end pairs.


- `-gil` <tex1t> <text2>...  
Labels for genomic intervals, space-separated list of strings, a separate value required for each gemomic intervals track.


- `-gic` <color1> [<color2>]  
Space-separated list of colors for genomic intervals tracks, uses the same logic as -bgc parameter.


- `-st` <transcript_id1> <transcript_id2>...  
A complete list of transcript IDs to show (others will be hidden, applicable to -w and -g modes).


- `-tl` <transcript_id> <label>  
Allows to replace the default <transcript_id> label with an arbitrary text <label>. 
For re-labeling several transcripts, several '-tl' keys can be used.

- `-tls` auto|both|id|name|gene_id|none  
Transcript label style. Can be 'none' (hide the transcript label), 'name', 'id' or 'gene_id' (show annotated transcript_name, transcript_id or gene_id respectively as a transcript_label), 'both' (show annotated transcript_name and transcript_id) and 'auto' (hides name if it is included in the ID)


- `-hrf` <N>  
Highlight a particular reading frame (+0, +1 or +2).


**MISCELLANEOUS ARGUMENTS**

- `-h`, `--help`  
Show this help message and exit.

- `-v`, `--version`  
Show program version.

- `--debug`  
Provide detailed stack trace for debugging purposes.

- `--verbose`  
Show all progress messages (default: False)

- `-hf` <x1> <x2> <label>  
Highlight and label a given segment of the genomic window (global coordinates on the selected contig).


- `-hc` <NtNtNt>  
Highlight selected codon on the aminoacid sequence track, e.g. -hc TGA.



### [Main page](https://github.com/art-egorov/svist4get)


#### [Quickstart guide](./QSGUIDE.md)

#### [Configuration file parameters](./CONFIG.md)

#### [API usage examples](./API.md)

#### [FAQ](./FAQ.md)

#### [Version log](./VERSION.md)

